src/x11.rs

```diff
@@ -131,33 +137,170 @@ impl X11Impl {
             .reply()
             .swbuf_err("Failed to get geometry reply")?;
 
+        // See if SHM is available.
+        let shm_info = {
+            let present = query_token
+                .reply()
+                .swbuf_err("Failed to get SHM query reply")?
+                .present;
+
+            if present {
+                // SHM is available.
+                Some(ShmInfo { seg: None })
+            } else {
+                None
+            }
+        };
+
         Ok(Self {
             connection,
             window,
             gc,
             depth: geometry_reply.depth,
+            shm: shm_info,
         })
     }
 
     pub(crate) unsafe fn set_buffer(&mut self, buffer: &[u32], width: u16, height: u16) {
         // Draw the image to the buffer.
-        let result = self.connection.put_image(
-            xproto::ImageFormat::Z_PIXMAP,
-            self.window,
-            self.gc,
-            width,
-            height,
-            0,
-            0,
-            0,
-            self.depth,
-            bytemuck::cast_slice(buffer),
-        );
-
-        match result {
-            Err(e) => log::error!("Failed to draw image to window: {}", e),
-            Ok(token) => token.ignore_error(),
+        let result = unsafe { self.set_buffer_shm(buffer, width, height) }.and_then(|had_shm| {
+            if had_shm {
+                Ok(())
+            } else {
+                log::debug!("Falling back to non-SHM method");
+                self.set_buffer_fallback(buffer, width, height)
+            }
+        });
+
+        if let Err(e) = result {
+            log::error!("Failed to draw image to window: {}", e);
+        }
+    }
+
+    /// Put the given buffer into the window using the SHM method.
+    ///
+    /// Returns `false` if SHM is not available.
+    ///
+    /// # Safety
+    ///
+    /// The buffer's length must be `width * height`.
+    unsafe fn set_buffer_shm(
+        &mut self,
+        buffer: &[u32],
+        width: u16,
+        height: u16,
+    ) -> Result<bool, PushBufferError> {
+        let shm_info = match self.shm {
+            Some(ref mut info) => info,
+            None => return Ok(false),
+        };
+
+        // Get the SHM segment to use.
+        let necessary_size = (width as usize) * (height as usize) * 4;
+        let (segment, segment_id) = shm_info.segment(&self.connection, necessary_size)?;
```

Comment:
Does anything make sure that the X11 server is done with reading from the shared memory before this writes to it again? (After all, this is the reason why `shm_put_image` can generate completion events via the `send_event` argument.)

Reply:
I originally had `send_event` sent to `true` and then used a loop that would call `wait_for_event` until it received a valid SHM completion event. However, I got paranoid that this loop would pick up and discard an actual valid input event that [`winit`](https://crates.io/crates/winit) might be concerned about.

Besides, [it doesn't look like `cairo` does it](https://github.com/freedesktop/cairo/blob/12cd2bcbb260f4ed64a72b64bd2fdd593f683bbc/src/cairo-xcb-connection-shm.c#L69-L71), although there might be some ameliorating control I missed while browsing `cairo`'s source code.

Reply:
Cairo does not control the main loop and thus cannot wait/poll for events. Instead, it (ab)uses the fact the X11 server handles requests in-order: It [sends a GetInputFocus request when it needs some syncronisation](https://github.com/freedesktop/cairo/blob/12cd2bcbb260f4ed64a72b64bd2fdd593f683bbc/src/cairo-xcb-shm.c#L303)  and then [uses `xcb_poll_for_reply()` to check whether that request is done](https://github.com/freedesktop/cairo/blob/12cd2bcbb260f4ed64a72b64bd2fdd593f683bbc/src/cairo-xcb-shm.c#L117).

Reply:
That sounds like a good strategy, I've implemented it. Besides, I believe it's also consistent with the fact that the Wayland backend blocks until the buffer's ready.